src/lib.rs

```diff
@@ -370,6 +408,28 @@ extern "C" fn fn_initialize(_init_args: CK_VOID_PTR) -> CK_RV {
         wstate.initialize();
     }
 
+    /* if a slot num was not specified, pick the first free one,
+     * but ensure that the same file is not already open on an
+     * existing slot */
+    if v.len() == 1 {
+        let mut slot_high: i64 = -1;
+        let slots = wstate.get_slots_ids();
+        for s in &slots {
+            if (*s) as i64 > slot_high {
+                slot_high = (*s) as i64;
+            }
+            match wstate.get_token_from_slot(*s) {
+                Ok(t) => {
+                    if filename.eq(t.get_filename()) {
+                        return CKR_CRYPTOKI_ALREADY_INITIALIZED;
+                    }
+                }
+                Err(_) => return CKR_GENERAL_ERROR,
+            }
+        }
+        slotnum = (slot_high + 1) as u64;
```

Comment:
what happens here if somebody creates a slot with ID UIN64_MAX value and then tries to initialize one more? Do we overflow here? Should we try to find first unused slot number? Could this have some side effects such as potential issues of slot ID reuse that could make login fails and token locked?

Reply:
NOTE: I was wrong on the above assertion, "as" does not crash at all, it instead does a binary cast and potentially a truncation if the receiving type is smaller.

But in this case it does not matter because if "slot_high +1" were to overflow we'd get a panic with error: "attempt to add with overflow" (just tested).